src/passes/Poppify.cpp

```diff
@@ -0,0 +1,504 @@
+/*
+ * Copyright 2021 WebAssembly Community Group participants
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+// Poppify.cpp - Transform Binaryen IR to Poppy IR.
+//
+// Poppy IR represents stack machine code using normal Binaryen IR types by
+// imposing the following constraints:
+//
+//  1. Function bodies and children of control flow (except If conditions) must
+//     be blocks.
+//
+//  2. Blocks may have any Expressions except for Pops as children. The sequence
```

Comment:
Once we support blocks taking values, I expect that more blocks will have pop children. Right now catch blocks are somewhat limited because this binary cannot be directly expressed in Binaryen or Poppy IR because our we need to have some instruction to consume the pop of the caught value:

```
try i32
  do
    i32.const 0
    throw $e ;; throws i32
  catch $e
    ;; i32 on stack
end
```

Eventually I would like to implement that catch with this Poppy IR block:

```
(block (params i32) (results i32)
  (pop i32)
)
```

Reply:
> Once we support blocks taking values, I expect that more blocks will have pop children.

Then this sounds like not an inherent characteristic of Poppy IR but a tentative restriction that will be removed later..? How about clarifying that in the comment?

> Right now catch blocks are somewhat limited because this binary cannot be directly expressed in Binaryen or Poppy IR because our we need to have some instruction to consume the pop of the caught value:
> 
> ```
> try i32
>   do
>     i32.const 0
>     throw $e ;; throws i32
>   catch $e
>     ;; i32 on stack
> end
> ```

I'm not sure if I understand what you mean by 'limited' and we have to 'consume' the caught value. Given that we don't support block inputs at the moment, `catch` seems to be an exception, but why is it limited? And I think it is possible not to consume the caught value if we flow it out of `try`, as in your example?

Reply:
I will update this comment to make clear that this restriction might be changed in the future.

I double checked and I was wrong about this example being invalid in Binaryen IR; sorry about that. There is no limitation in Binaryen IR after all. Right now the validator does improperly reject the Poppy IR version of that example, but I will fix that in a follow-up along with other validation issues with Poppy IR.